File: system-test/java/src/main/java/com/optivem/shop/testkit/dsl/core/usecase/shop/usecases/PublishCoupon.java
package com.optivem.shop.testkit.dsl.core.usecase.shop.usecases;

import com.optivem.shop.testkit.driver.port.shop.dtos.PublishCouponRequest;
import com.optivem.shop.testkit.driver.port.shop.ShopDriver;
import com.optivem.shop.testkit.dsl.core.usecase.shop.usecases.base.BaseShopUseCase;
import com.optivem.shop.testkit.dsl.core.shared.UseCaseResult;
import com.optivem.shop.testkit.dsl.core.shared.UseCaseContext;
import com.optivem.shop.testkit.dsl.core.shared.VoidVerification;

public class PublishCoupon extends BaseShopUseCase<Void, VoidVerification> {
    private String couponCodeParamAlias;
    private String discountRate;
    private String validFrom;
    private String validTo;
    private String usageLimit;

    public PublishCoupon(ShopDriver driver, UseCaseContext context) {
        super(driver, context);
    }

    public PublishCoupon couponCode(String couponCodeParamAlias) {
        this.couponCodeParamAlias = couponCodeParamAlias;
        return this;
    }

    public PublishCoupon discountRate(String discountRate) {
        this.discountRate = discountRate;
        return this;
    }

    public PublishCoupon validFrom(String validFrom) {
        this.validFrom = validFrom;
        return this;
    }

    public PublishCoupon validTo(String validTo) {
        this.validTo = validTo;
        return this;
    }

    public PublishCoupon usageLimit(String usageLimit) {
        this.usageLimit = usageLimit;
        return this;
    }

    @Override
    public UseCaseResult<Void, VoidVerification> execute() {
        var couponCode = context.getParamValue(couponCodeParamAlias);

        var request = PublishCouponRequest.builder()
                .code(couponCode)
                .discountRate(discountRate)
                .validFrom(validFrom)
                .validTo(validTo)
                .usageLimit(usageLimit)
                .build();

        var result = driver.publishCoupon(request);
        return new UseCaseResult<>(result, context, VoidVerification::new);
    }
}

File: system-test/typescript/src/testkit/dsl/core/usecase/shop/usecases/PublishCoupon.ts
import type { ShopDriver } from '../../../../../driver/port/shop/shop-driver.js';
import type { PublishCouponRequest } from '../../../../../driver/port/shop/dtos/PublishCouponRequest.js';
import { UseCaseResult } from '../../../shared/use-case-result.js';
import { VoidVerification } from '../../../shared/void-verification.js';
import type { UseCaseContext } from '../../../shared/use-case-context.js';
import { BaseShopUseCase } from './base/BaseShopUseCase.js';

export class PublishCoupon extends BaseShopUseCase<void, VoidVerification> {
  private _couponCodeParamAlias?: string;
  private _discountRate: number | string = '';
  private _validFrom?: string;
  private _validTo?: string;
  private _usageLimit?: number | string;

  constructor(driver: ShopDriver, context: UseCaseContext) {
    super(driver, context);
  }

  couponCode(couponCodeParamAlias: string): this {
    this._couponCodeParamAlias = couponCodeParamAlias;
    return this;
  }

  discountRate(discountRate: number | string): this {
    this._discountRate = discountRate;
    return this;
  }

  validFrom(validFrom: string): this {
    this._validFrom = validFrom;
    return this;
  }

  validTo(validTo: string): this {
    this._validTo = validTo;
    return this;
  }

  usageLimit(usageLimit: number | string): this {
    this._usageLimit = usageLimit;
    return this;
  }

  async execute(): Promise<UseCaseResult<void, VoidVerification>> {
    const couponCode = this.context.getParamValue(this._couponCodeParamAlias);

    const request: PublishCouponRequest = {
      code: couponCode ?? '',
      discountRate: this._discountRate,
      validFrom: this._validFrom,
      validTo: this._validTo,
      usageLimit: this._usageLimit,
    };

    const result = await this.driver.publishCoupon(request);

    return new UseCaseResult(
      result,
      this.context,
      (_, ctx) => new VoidVerification(undefined, ctx),
    );
  }
}
